File: pressure.cpp
#include "pressure.h"

using namespace std;

/**************************************************
* Constructor Method
Creates a Default Sensor
-----------------------------------
Address = 0x28
Bus = /dev/i2c-1
Pressure Span = 0-100 mBar
Min ADC counts = ADC_MIN_COUNTS
Max ADC counts = ADC_MAX_COUNTS
-----------------------------------
Args: I2C Bus Number
Returns: New Pressure Sensor Object
***************************************************/

Pressure::Pressure(){
	setBus(1);
	setAddress(DEFAULT_ADDRESS);
	setSpan(0,100);
	setCounts(ADC_MIN_COUNTS,ADC_MAX_COUNTS);
	getFd();
}

/**************************************************
* Constructor Method
Creates a Default Sensor, I2C Bus Declared
-----------------------------------
Address = 0x28
Bus = user defined
Pressure Span = 0-100 mBar
Min ADC counts = ADC_MIN_COUNTS
Max ADC counts = ADC_MAX_COUNTS
-----------------------------------
Args: I2C Bus Number
Returns: New Pressure Sensor Object
***************************************************/

Pressure::Pressure(int myBus){
	setBus(myBus);
	setAddress(DEFAULT_ADDRESS);
	setSpan(0,100);
	setCounts(ADC_MIN_COUNTS,ADC_MAX_COUNTS);
	getFd();
}

/**************************************************
* Constructor Method
Creates a Default Sensor, I2C Bus Declared, user-defined address
-----------------------------------
Pressure Span = 0-100 mBar
Min ADC counts = ADC_MIN_COUNTS
Max ADC counts = ADC_MAX_COUNTS
-----------------------------------
Args: I2C Bus Number
Returns: New Pressure Sensor Object
***************************************************/

Pressure::Pressure(int myBus, int myAddress){
	setBus(myBus);
	setAddress(myAddress);
	setSpan(0,100);
	setCounts(ADC_MIN_COUNTS,ADC_MAX_COUNTS);
	getFd();
}
/**************************************************
* Constructor Method
Creates a Default Sensor, I2C Bus Declared, user-defined address, User Defined Span)
-----------------------------------
Min ADC counts = ADC_MIN_COUNTS
Max ADC counts = ADC_MAX_COUNTS
-----------------------------------
Args: I2C Bus Number
Returns: New Pressure Sensor Object
***************************************************/
Pressure::Pressure(int myBus, int myAddress,int myMin, int myMax){
	setBus(myBus);
	setAddress(myAddress);
	setSpan(myMin,myMax);
	setCounts(ADC_MIN_COUNTS,ADC_MAX_COUNTS);
	getFd();
}


/**************************************************
* Function counts2mbar()
Descritpion: Converts Pressure Transducer ADC Counts
to pressure in millibars.
Args: counts - Transducer ADC Counts
Returns: pressure - Pressure in Milibars
***************************************************/
float Pressure::counts2mbar(int counts){

	float myPressure;

	myPressure=((((float)counts-countsmin)*(pmax-pmin))/((float)countsmax-(float)countsmin))+pmin;
	return myPressure;
}
/**************************************************
* Function counts2mmhg()
Descritpion: Converts Pressure Transducer ADC Counts
to pressure in mmhg.
Args: counts - Transducer ADC Counts
Returns: pressure - Pressure in mmHg
***************************************************/
float Pressure::counts2mmhg(int counts){

	return (counts2mbar(counts)*0.7500);
}

/**************************************************
* Function mbar2mmhg()
Descritpion: Converts Pressure Transducer ADC Counts
to pressure in millibars.
Args: p_mbar -
Returns: pressure in mmHg
***************************************************/
float Pressure::mbar2mmhg(float p_mbar){

	return p_mbar*0.7500;  // .75 factor conversion
}
/**************************************************
* Function setAddress()
Descritpion: Sets the I2C Address of the pressure sensor
Args: myAddress - I2C Address
Returns: None
***************************************************/
void Pressure::setAddress(int myAddress){
	address=myAddress;
}
/**************************************************
* Function setBus(int)
Descritpion: Sets the I2C Bus of the pressure sensor
Args: myBus - I2C Bus
Returns: None
***************************************************/
void Pressure::setBus(int myBus){
	bus=myBus;
}
/**************************************************
* Function setSpan(int,int)
Descritpion: Sets the Span of the pressure sensor
Args: min - Min Value
      max - Max Value
Returns: None
***************************************************/
void Pressure::setSpan(int min, int max){
	pmin=min;
	pmax=max;
}
/**************************************************
* Function setFd(int)
Descritpion: Sets the Transducer Format Descirptor
Value
Args: fd - format desciriptor integer
Returns: None
***************************************************/
void Pressure::getFd(){

	if(bus==0){
	fd = wiringPiI2CSetupInterface("/dev/i2c-0", address);
	printw("format descriptor: %i\n",fd);
	}
	else{
	fd = wiringPiI2CSetupInterface("/dev/i2c-1", address);
	printw("format descriptor: %i\n",fd);
	}

}
/**************************************************
* Function setFd(int)
Descritpion: Sets the Transducer Format Descirptor
Value
Args: fd - format desciriptor integer
Returns: None
***************************************************/
void Pressure::setFd(int myFd){
	fd=myFd;
}
/**************************************************
* Function setCounts()
Descritpion: Sets the Min and Max ADC Counts
of the pressure sensor
Args: min - Min Value
      max - Max Value
Returns: None
***************************************************/
void Pressure::setCounts(int min, int max){
	countsmin=min;
	countsmax=max;
}
/**************************************************
* Function readCounts()
Descritpion: Reads the current pressure sensor data
register.
Returns: Presure Sensor Current ADC Counts
***************************************************/
int Pressure::readCounts(){
int data, adCounts;
	data=wiringPiI2CReadReg16(fd,DATA_REG); //read 16 bits from the data register

	/*the 2 bytes that come out of the data register are reversed(Little Endian).
	so i flip them here by doing a bitwise AND for each byte, bitshfting them each 8 places
	and then OR them together to get the data register in the correct order*/
	adCounts=((data & 0x00FF) << 8) | ((data & 0xFF00) >> 8);
	return adCounts;
}
/**************************************************
* Function readPressure_mmhg()
Descritpion: Reads the current pressure sensor data
register and performs the ADC to mmHg conversion.

Returns: Presure Sensor Current Value in mmHg
***************************************************/
float Pressure::readPressure_mmhg(){
int data;
	data=readCounts();
	return (counts2mmhg(data));
}
/**************************************************
* Function readPressure_mmhg()
Descritpion: Reads the current pressure sensor data
register and performs the ADC to mBar conversion.

Returns: Presure Sensor Current Value in mBar
***************************************************/
float Pressure::readPressure_mbar(){
int data;
	data=readCounts();
	return (counts2mbar(data));
}

File: go_src/pressure.go
package main

import (
	"bytes"
	"fmt"
	"strconv"
	"time"

	"golang.org/x/exp/io/i2c"
)

// Constants
const (
	// mBar to mmHg conversion Ratio
	MMHG_MBAR_RATIO = 0.7500

	// Define Device min and max ADC counts
	ADC_MIN_COUNTS = 0x0666
	ADC_MAX_COUNTS = 0x3999

	// Define Device Data Register Value
	DATA_REG             = 0x00
	DEFAULT_ADDR         = 0x28
	DEFAULT_MAX_PRESSURE = 100
	DEFAULT_MIN_PRESSURE = 0
	DEFAULT_BUS          = 0
	DEFAULT_DEVICE       = "/dev/i2c-0"
)

// Class Declration
type Pressure struct {
	bus       int        // the i2c Bus of the sensor
	address   int        // i2c Address of the sensor
	minCounts int        // Minimum Counts (Zero Reading) for the sensor
	maxCounts int        // Maximum Counts (Full Scale Reading) for the sensor
	pMin      int        // Minimum Pressure (Zero Reading) for the sensor
	pMax      int        // Maximum Pressure (Full Scale Reading) for the sensor
	i2cHandle uint32     // Reference to the I2C Handle
	data      float32    // Stored Data
	databus   i2c.Device //Create an I2C Device Variable
}

//Setter Functions

/*
*************************************************
Function set_bus(int)
Descritpion: Sets the I2C Bus of the pressure sensor
Args: myBus - I2C Bus
Returns: None
**************************************************
*/
func (sensor *Pressure) set_bus(myBus int) {
	sensor.bus = myBus
}

/*
*************************************************
Function set_address(int)
Descritpion: Sets the I2C Address of the pressure sensor
Args: myAddr - I2C Address
Returns: None
**************************************************
*/
func (sensor *Pressure) set_address(myAddr int) {
	sensor.address = myAddr
}

/*
*************************************************
Function set_min_counts(int)
Descritpion: Sets the Minumum Counts of pressure sensor
Args: counts- min ADC Counts from Sensor Data Sheet.
Returns: None
**************************************************
*/
func (sensor *Pressure) set_min_counts(myCounts int) {
	sensor.minCounts = myCounts
}

/*
*************************************************
Function set_max_counts(int)
Descritpion: Sets the maximum counts of pressure sensor
Args: counts- max ADC Counts from Sensor Data Sheet.
Returns: None
**************************************************
*/
func (sensor *Pressure) set_max_counts(myCounts int) {
	sensor.maxCounts = myCounts
}

/*
*************************************************
Function set_max_pressure(int)
Descritpion: Sets the maximum pressure range of pressure sensor
in mBar
Args: counts- max pressure from Sensor Data Sheet.
Returns: None
**************************************************
*/
func (sensor *Pressure) set_max_pressure(myPressure int) {
	sensor.pMax = myPressure
}

/*
*************************************************
Function set_max_pressure(int)
Descritpion: Sets the maximum pressure range of pressure sensor
in mBar
Args: counts- max pressure from Sensor Data Sheet.
Returns: None
**************************************************
*/
func (sensor *Pressure) set_min_pressure(myPressure int) {
	sensor.pMin = myPressure
}

// Initialize I2C Bus, this probably needs to be broken up better
// Need to make this so that the default bus and device can change
func (sensor *Pressure) initialize() {

	//	logger.ChangePackageLogLevel("i2c", logger.InfoLevel)
	//	i2cdevice, err := i2c.NewI2C(uint8(sensor.address), sensor.bus)

	i2cdevice, err := i2c.Open(&i2c.Devfs{Dev: "/dev/i2c-1"}, DEFAULT_ADDR)
	if err != nil {
		panic(err)
	}

	sensor.databus = *i2cdevice
}

/*
*************************************************
Function readCounts()
Descritpion: Reads the current pressure sensor data
register.
Returns: Presure Sensor Current ADC Counts
**************************************************
*/
func (sensor *Pressure) read_counts() int {

	//	dataSample, err := sensor.databus.ReadRegU16BE(DATA_REG)
	var buffer []byte
	buffer = append(buffer, 2)
	dataBuffer := sensor.databus.Read(buffer)
	dataSample := bytes.Buffer(dataBuffer)
	return int(dataSample)
}

/*
*************************************************
Function read_mbar()
Descritpion: Reads the current pressure sensor data
and converts it to milibars
Returns: Presure Sensor Pressure in Milibars (Float32)
**************************************************
*/
func (sensor *Pressure) read_mbar() float32 {

	return (sensor.counts2mbar(sensor.read_counts()))

}

/*
*************************************************
Function read_mbar()
Descritpion: Reads the current pressure sensor data
and converts it to milibars
Returns: Presure Sensor Pressure in Milibars (Float32)
**************************************************
*/
func (sensor *Pressure) read_mmhg() float32 {

	return (sensor.read_mbar() * MMHG_MBAR_RATIO)

}

/*
*************************************************
Function counts2mbar()
Descritpion: Converts Counts to Milibar Pressure
Returns: pMbar - Pressure in Milibars (Float32)
**************************************************
*/
func (sensor *Pressure) counts2mbar(counts int) float32 {

	var pMbar float32
	num := float32(int(counts)-sensor.minCounts) * float32(sensor.pMax-sensor.pMin) // Numerator
	den := float32(sensor.maxCounts - sensor.minCounts)                             //Denominator
	pMbar = (num / den) + float32(sensor.pMin)                                      //add min pressure, should add an additional calibration offset.
	return pMbar
}

func newSensor() *Pressure {
	sensor := Pressure{}
	sensor.set_address(DEFAULT_ADDR)
	sensor.set_bus(DEFAULT_BUS)
	sensor.set_max_counts(ADC_MAX_COUNTS)
	sensor.set_min_counts(ADC_MIN_COUNTS)
	sensor.set_max_pressure(DEFAULT_MAX_PRESSURE)
	sensor.set_min_pressure(DEFAULT_MIN_PRESSURE)

	sensor.initialize() // initialize Sensor

	return &sensor
}

func main() {

	// Declare a Sensor Object
	var sensor1 = newSensor()
	fmt.Println("Sensor Bus : ", sensor1.bus)
	var count = 0
	for i := 0; i < 100; i++ {

		dataSample := sensor1.read_counts()
		fmt.Println("mmHg Data Read :", strconv.FormatInt(int64(dataSample), 16))
		time.Sleep(100 * time.Millisecond)
		if dataSample > 2000.0 {
			count++
		}
	}
	println("error count: ", count)
	sensor1.databus.Close()
}
